### ISSUE TITLE
Gradient of gather uses 2x to 4x memory

### ISSUE THREAD

**Issue by Findus23** [jax-ml/jax#30015]
### Description

I have a simple gather operation `x[tuples]` I want to apply autodiff to. But using `jax.grad` on the function uses 2x or 4x as much memory as the original function.
This might be a bug in XLA instead of JAX, but I am reporting it here as it happens easily using JAX.

This is a simple example showing the issue:
```python
import os
import sys
from pathlib import Path

import jax
import jax.numpy as jnp

grads = sys.argv[1] == "grads"
mode = "grads" if grads else "nograds"

out_dir = f"tmp_dump/{mode}"

os.environ["XLA_FLAGS"] = f"--xla_dump_to={out_dir}"

idx = jax.ShapeDtypeStruct((2048 * 2048 * 2048, 3), jnp.int32) # or jnp.int16

x = jax.ShapeDtypeStruct((128, 128, 128), jnp.float32)

def function_using_idx(x, idx):
    tuples = tuple(jnp.split(idx, indices_or_sections=3, axis=-1))
    gathered = x[tuples][..., 0]
    return jnp.sum(gathered * 2.5)

if grads:
    function_using_idx = jax.grad(function_using_idx, argnums=0)

hlo = jax.jit(function_using_idx).lower(x, idx).compile()

mem_report = Path(out_dir) / "module_0001.jit_function_using_idx.cpu_after_optimizations-memory-usage-report.txt"

with mem_report.open() as f:
    print(next(f),end="")
```
```
 ➜  ~ python trivial_idx.py nograds
Total bytes used: 138521083996 (129.01GiB)
 ➜  ~ python trivial_idx.py grads
Total bytes used: 240526557204 (224.01GiB)
```
Using those array sizes, the function input `idx` (`s32[8589934592,3]`) is 96GB large and the intermediate `gathered` (`f32[8589934592]`) is 32GB large. The regular jitted function can run with only 1GB additional memory at the peak (data from `buffer-assignment.txt`:

| Cumulative Size | Size     | Name                          | op_name                               | source            | array_info        | alloc |
|-----------------|----------|-------------------------------|---------------------------------------|-------------------|-------------------|----|
| 96.0 GiB        | 96.0 GiB | Arg_1.2                       | idx                                   |                   | s32[8589934592,3] | 0  |
| 128.0 GiB       | 32.0 GiB | multiply_bitcast_fusion.clone | jit(function_using_idx)/jit(main)/mul | trivial_idx.py:16 | f32[8589934592]   | 25 |
| 129.0 GiB       | 1.0 GiB  | reduce-window.clone           |                                       |                   | f32[268435456]    | 25 |
| 129.0 GiB       | 8.0 MiB  | Arg_0.1                       | x                                     |                   | f32[128,128,128]  | 1  |

But in comparison to that when using `jax.grad`, the peak memory usage increases dramatically as the input array gets copied into a second buffer of the same shape:

| Cumulative Size | Size     | Name                             | op_name                                   | source            | array_info            | alloc |
|-----------------|----------|----------------------------------|-------------------------------------------|-------------------|-----------------------|---|
| 96.0 GiB        | 96.0 GiB | Arg_0.1                          | idx                                       |                   | s32[8589934592,3]     | 0 |
| 192.0 GiB       | 96.0 GiB | concatenate_bitcast_fusion.clone | jit(function_using_idx)/jit(main)/concat… | trivial_idx.py:15 | s32[8589934592,3]     | 7 |
| 224.0 GiB       | 32.0 GiB | broadcast.4.clone                | jit(function_using_idx)/jit(main)/broadc… | trivial_idx.py:15 | f32[8589934592,1,1,1] | 7 |
| 224.0 GiB       | 8.0 MiB  | broadcast.3.clone                |                                           |                   | f32[128,128,128]      | 1 |
| 224.0 GiB       | 8.0 MiB  | wrapped_scatter.clone            | jit(function_using_idx)/jit(main)/scatte… | trivial_idx.py:15 | f32[128,128,128]      | 1 |

This gets even worse when using jnp.int16 for `idx` as the copy will still be `s32` (therefore taking up 4x as much memory as without that copy).

Checking the HLO it seems to me like the duplicate array is not needed in any place:
```mlir
[...]
%parallel_concatenate_bitcast_fusion (p.2: s32[8589934592,3]) -> s32[8589934592,3] {
  %p.2 = s32[8589934592,3]{1,0} parameter(0)
  ROOT %concatenate_bitcast_fusion.clone = s32[8589934592,3]{1,0} fusion(%p.2), kind=kLoop, calls=%fused_computation.clone, metadata={op_name="jit(function_using_idx)/jit(main)/concatenate" source_file=".../trivial_idx.py" source_line=15}, backend_config={"outer_dimension_partitions":["4"]}
}
[...]
ENTRY %main.38 (Arg_0.1: s32[8589934592,3]) -> f32[128,128,128] {
  %Arg_0.1 = s32[8589934592,3]{1,0} parameter(0), metadata={op_name="idx"}
  %constant.2 = f32[] constant(0)
  %constant.4 = f32[] constant(2.5)
  %call.2 = s32[8589934592,3]{1,0} call(%Arg_0.1), to_apply=%parallel_concatenate_bitcast_fusion
  %call = f32[128,128,128]{2,1,0} call(%constant.2), to_apply=%parallel_broadcast.3
  %call.1 = f32[8589934592,1,1,1]{3,2,1,0} call(%constant.4), to_apply=%parallel_broadcast.4
  ROOT %call.3 = f32[128,128,128]{2,1,0} call(%call, %call.2, %call.1), to_apply=%parallel_wrapped_scatter
}
```

Some more thoughts:

- This seems to be unrelated to 3D. Even a simplified 1D version has the same issue:
```python
idx = jax.ShapeDtypeStruct((2048 * 2048 * 2048,), jnp.int32) # or jnp.int16
x = jax.ShapeDtypeStruct((128,), jnp.float32)
def function_using_idx(x, idx):
    gathered = x[idx]
    return jnp.sum(gathered * 2.5)
```
- adding `--xla_disable_hlo_passes='fusion'` causes both the grad and non-grad version to use the same (even larger) amount of memory

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.6.2
jaxlib: 0.6.2
numpy:  2.3.1
python: 3.13.3 (main, Apr 10 2025, 21:38:51) [GCC 14.2.0]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='lukasnotebook', release='6.12.32-amd64', version='#1 SMP PREEMPT_DYNAMIC Debian 6.12.32-1 (2025-06-07)', machine='x86_64')
```

**Comment by Findus23** [comment]
After testing the above on CPU, I now quickly checked on a device with a CUDA GPU and surprisingly enough there the grad function seems to be using the same amount of memory. Also weirdly enough checking `memory-usage-report.txt` in both cases there seems to be only one >10MB memory allocation which is the 96GB `idx`.
So I am unsure if this is a good method to debug memory usage of complex programs in a simple test-case.

Also XLA prints
```
2025-07-07 19:45:04.661281: E external/xla/xla/service/gpu/gpu_hlo_schedule.cc:795] The byte size of input/output arguments (103087603712) exceeds the base limit (19001278464). This indicates an error in the calculation!
2025-07-07 19:45:04.661859: W external/xla/xla/hlo/transforms/simplifiers/hlo_rematerialization.cc:3023] Can't reduce memory use below 0B (0 bytes) by rematerialization; only reduced to 16.00MiB (16777216 bytes), down from 16.00MiB (16777216 bytes) originally
```
